### DISCUSSION TITLE
How to Proxy Multiple Code-Servers via Nginx?

### DISCUSSION THREAD

**Discussion by yuzhuoheng** [coder/code-server#7036] [Q&A] [answered]
I have multiple code-servers deployed in Docker containers. They are located at 127.0.0.1:13001, 127.0.0.1:13002, and so on. How can I achieve this through Nginx? For example, accessing /code/13001 should forward to 127.0.0.1:13001, and accessing /code/13002 should forward to 127.0.0.1:13002. It's important to note that the ports I mentioned (e.g., 13001) are not fixed.

I tried using the following configuration, but the browser cannot open the page normally.
![企业微信截图_17288923142006](https://github.com/user-attachments/assets/878a1978-85df-45ac-b299-e9c6820b05b0)
![1728892397496](https://github.com/user-attachments/assets/13aac600-edf3-47af-bdb0-83c93105ddba)
![1728892417464](https://github.com/user-attachments/assets/4714700e-9a04-43de-a3d1-f83b09d0d170)
![1728892443648](https://github.com/user-attachments/assets/27548dfc-7de1-463f-ac02-b23890eb919b)

**Comment by code-asher** [comment]
I think once you use a regex you have to tell NGINX what the full target is, but I am not sure.  This seems to work for me:

```
proxy_pass http://127.0.0.1:$1$2$is_args$args;
```

**Reply by code-asher** [reply]
Or you can probably use `rewrite` as well.

**Reply by yuzhuoheng** [reply]
Thank you very much for being willing to help me!! But I still haven't solved it。

I have modified nginx according to your suggestion and hope to use/vscode/31013/proxy to access it http://127.0.0.1:31013
![企业微信截图_17291354477082](https://github.com/user-attachments/assets/a9ffa434-f1ee-4c17-8cee-9945bbb05062)


After the nginx - s reload, I accessed it using a browser http://192.168.8.74:31013/ , can be accessed normally
![企业微信截图_17291354702664](https://github.com/user-attachments/assets/613f5ad1-81cb-42b8-a983-5c19472b183f)


But I use https://vscode.rd.supcon.com/vscode/31013/ The page displays' not found '
![企业微信截图_17291354936262](https://github.com/user-attachments/assets/1f4f78d6-cc66-4639-bf10-9f5f665be034)


To verify if there is an issue with nginx, I will follow your answers in other posts. Added a section of ngixn configuration:
![企业微信截图_17291355435666](https://github.com/user-attachments/assets/6ba73c6e-3252-4a1c-8fe4-1479b749870f)


After nginx - s reload, access it again through the browser: https://vscode.rd.supcon.com/code/ , can be accessed normally
![企业微信截图_17291355703776](https://github.com/user-attachments/assets/fb887c61-f11d-40e3-9370-03ee4391fbc7)

**Reply by code-asher** [reply]
Your `location` is a little different than what you originally posted.  This is what I was using (notice the `$2`):

```
location ~/code/(\d+)(/.*)?$ {
  proxy_pass http://127.0.0.1:$1$2$is_args$args;
  proxy_set_header ... etc etc ....
```

But, I was able to make `~ ^/vscode/(.*)/?$` work by omitting the slash in `proxy_pass`:

```
location ~ ^/vscode/(.*)/?$ { 
  proxy_pass http://127.0.0.1:$1$is_args$args; 
  proxy_set_header ... etc etc ....
```

**Reply by yuzhuoheng** [reply]
I used your configuration:
![企业微信截图_17364973836710](https://github.com/user-attachments/assets/32164fa9-0f07-4b52-be36-5eb775b27d11)

When I opened the following URL in my browser: https://hos-studio.rd.supcon.com/vscode/15555/

I noticed that the page resources loaded correctly, but its WebSocket was not properly forwarded.
![企业微信截图_17364976371527](https://github.com/user-attachments/assets/41a5a67b-88ff-4fd1-b593-5550e5900578)

**Reply by code-asher** [reply, marked as answer]
Make sure to include all the other headers as well.  NGINX will not forward web sockets without `Upgrade` and `Connection`.  I would include `Host` and `Accept-Encoding` as well.

**Comment by BigFaceBoy** [comment]
@yuzhuoheng 这个问题你解决了吗
